added script automatically creating ../src/CbindWrapper.f95, which is…

rainbowsend opened the pull request:
closes #177 

added an adapted version of [src/create_wrapper.py](https://github.com/SHTOOLS/SHTOOLS/blob/master/src/create_wrapper.py) named [src/create_cbind_wrapper.py](https://github.com/rainbowsend/SHTOOLS/blob/cpp/src/create_cbind_wrapper.py) 

Automatically creates wrapper with iso_c_binding
- replacing assumed size arrays with fixed size arrays
  - additional arguments for each dimension. Dimension arguments follow the corresponding array argument in the signature
  - special case for variables called `cilm`: only one additional dimension is added. I assume first dimension is always 2 and second and third dimension are equal
  - there are more cases where dimension size is known during compile time, but this can not be considered automatically.
- replacing Fortran types with c types
- names are in snake_case. camelCase would be consistent with Fortan names, though.

Example:

```Fortran
function cbindMakeGridPoint(cilm,cilm_d,lmax,lat,lon,norm,csphase,dealloc)  bind(c&
                                    , name="cbind_make_grid_point")
        use, intrinsic :: iso_c_binding
        use shtools, only: MakeGridPoint
        implicit none
        real(kind=c_double) :: cbindMakeGridPoint
        integer(kind=c_int), intent(in) :: cilm_d
        real(kind=c_double), dimension(2,cilm_d,cilm_d),intent(in) :: cilm
        real(kind=c_double), intent(in) :: lat
        real(kind=c_double), intent(in) :: lon
        integer(kind=c_int), intent(in) :: lmax
        integer(kind=c_int), intent(in) :: norm
        integer(kind=c_int), intent(in) :: csphase
        integer(kind=c_int), intent(in) :: dealloc
        cbindMakeGridPoint=MakeGridPoint(cilm,lmax,lat,lon,norm=norm,csphase=csphase&
                                             ,dealloc=dealloc)
end function cbindMakeGridPoint
```
Complete wrapper file including all subroutines: [src/CbindWrapper.f95](https://github.com/rainbowsend/SHTOOLS/blob/cpp/src/CbindWrapper.f95)


In cpp the wrapper can be used to call the Fortan routines. I have tested this for some [routines](https://github.com/rainbowsend/SHTOOLS/tree/master/cpp/src). These cpp test are not part of this PR. You can find some extract below

```cpp

#pragma once

extern "C"
{ 
     double cbind_make_grid_point( const double* cilm, const int* cilm_d0,
                                   const int* lmax,
                                   const double* lat, const double* lon,
                                   const int* norm,
                                   const int* csphase,
                                   const int* dealloc);

}

inline double cpp_make_grid_point( const std::vector<double>& cilm, double lat, double lon, 
                                     int norm=1, int csphase=1, int dealloc=0)
{

   int n = cilm.size();
   int cilmd = std::sqrt(n/2);
   int lmax = cilmd-1;
      
   return cbind_make_grid_point (&cilm[0], &cilmd, &lmax, &lat, &lon, &norm, &csphase, &dealloc);

 }
```
For each subroutine/function in  [src/CbindWrapper.f95](https://github.com/rainbowsend/SHTOOLS/blob/cpp/src/CbindWrapper.f95) we need a corresponding function in the `extern "C"` block. I have added three so far for test purposes. But in the long term they should be created automatically with some script.

```cpp

int
main(int argc, char** argv)
{
  std::string infile = "../examples/ExampleDataFiles/MarsTopo719.shape";

  int lmax = 15;
  int n = lmax+1; 
  std::vector<double> mars = cpp_sh_read(infile, lmax );
      
  double val = cpp_make_grid_point( mars, 10.0, 30.0);
  std::cout <<  std::setprecision(16) << val << std::endl;
  std::cout <<  "diff to python " << val-3395259.548270001 << std::endl;
  
  return 0;
}

```

In this PR src/CbindWrapper.f95 is not in the Makefile

MarkWieczorek commented:
Could you let me know how this deals with the Fortran 95 optional arguments? Do you need to supply all arguments to these routines? Or are you allowed to leave some argument unspecified? (Sorry, but its been a long time since I used c.)

The reason I ask is that the Fortran routines check to see which optional parameters have been specified, and in some cases, not all are allowed. For example, with the routine `SHMultiTaperSE`, you can either specify `lat` and `lon`, or `alpha`, but not all three. When writing the python wrappers, we had to make some choices as to how to deal with this, and if needed we could do the same here.

Similar issues arise with (and I might have missed some)
```
SHMultiTaperCSE
SHLocalizedAdmitCorr
SHExpandGLQ
SHExpandGLQC
MakeGridGLQC
MakeGridGLQ
CilmPlusRhoH
CilmPlus
CilmMinusRho
CilmMinus
BAtoHilmRhoH
BAtoHilm
```

rainbowsend commented:
you can simply pass a null pointer for optional arguments. This also works for arrays, even if the additional dimension argument is not zero.

To illustrate this if have changed the  `cpp_make_grid_point` from above

```cpp
inline double cpp_make_grid_point( const std::vector<double>& cilm, double lat, double lon )
{
   int n = cilm.size();
   int cilmd = std::sqrt(n/2);
   int lmax = cilmd-1;
   
   int* norm=nullptr;
   int* csphase=nullptr;
   int* dealloc=nullptr;
      
   return cbind_make_grid_point (&cilm[0], &cilmd, &lmax, &lat, &lon, norm, csphase, dealloc);
 }
```
The optional flag is missing in src/CbindWrapper.f95 in this version.  In a new version I have added them, but it is not making a difference.

rainbowsend commented:
apparently the `optional` keyword can cause problems:

with GNU Fortran (SUSE Linux) 7.5.0 I get  error messages like this

```
Error: TS 29113/TS 18508: Variable ‘exitstatus’ at (1) with OPTIONAL attribute in procedure ‘cshslepianvar’ which is BIND(C)
cWrapper.f95:2494:90:
```
for every argument declared `optional`

on another machine with
GNU Fortran (GCC) 8.4.0 or GNU Fortran (GCC) 10.2.0
everything compiles without error

A solution is using `-std=f2008ts` flag with gfortran 7.5

MarkWieczorek commented:
I'm hoping to push out a new release in a few days (I'm just waiting for pygmt to finish ta release that is in prgress), so I'm curious what more needs to be done before merging this.

I see a couple things:

* In this PR, it looks like src/cWrapper.f95 is no longer compiled in the makefile, so I guess this should be added.
* The main Makefile uses gfortran with `-std=f2003` by default. I've tested `-std=f2008` and `-std=f2008ts`, and they seem to work ok on my mac. I don't understand what the `ts` in `f2008ts` means. Do you think I should change `-std` to one of these?
* I think that it would be great to include a working example in the folder `examples/cpp`. I doesn't matter to me how simple the example is: I just want to be able to test that the cbindings are working, in principle. I will probably add this test to travis to make sure nothing breaks with future releases.

rainbowsend commented:
> https://gcc.gnu.org/onlinedocs/gfortran/Fortran-Dialect-Options.html
> The deprecated option ‘-std=f2008ts’ acts as an alias for ‘-std=f2018’. It is only present for backwards compatibility with earlier gfortran versions and should not be used any more. 

to compile cWrapper.f95 the newest fortran standard is required. ‘-std=f2018' was added in gfortran version 8. Thus, for gfortran version 7 ‘-std=f2008ts’ is needed. 
Maybe compiling cWrapper.f95 should be optional, so users not interested in c compatibility, do not depend on fortran 2018 and still can use  ‘-std=f2003'

I will create some small example without external dependencies

MarkWieczorek commented:
Would `-std=gnu` solve the problem for version 7?

I don't really care which standard we use. If I remember correctly, I choose 2003 only because this was the minimum necessary for iso_c_binding and/or linking to the fftw cbinded library.

rainbowsend commented:
`-std=gnu` works at least for me on both version 7 and 10

MarkWieczorek commented:
Why don't we just change it to gnu then. I'll just change it back manually if I ever have to test strict adherence to a certain standard (which I probably won't have to).

rainbowsend commented:
I am going to change the wrapper, so that it uses  fixed dimension whenever possible. However, this can only be done manually. Hope I am fast enough.
Once this is done, another nice thing would a documentation for `shtools.h`

MarkWieczorek commented:
I will release v4.7 on monday/tuesday once I get back to the lab. If there is anything here that is ready to go, let me know before then.

rainbowsend commented:
removed leading 'c' for function names in shtools.h

MarkWieczorek commented:
fyi, I am probably going to put out a maintenance release (v4.7.1) next week some time: https://github.com/SHTOOLS/SHTOOLS/pull/253. This is mostly to update a few minor issues with the makefiles that are stopping me from getting the fortran library accepted at homebrew-core and macports.

Is this about ready to go? One way of dealing with this would be to include this with 4.7.1, and add the "documentation" later. This would allow a few people to test this on the released version before advertising that is 100% ready.

rainbowsend commented:
That is a good approach. The wrappers are finished, so it is ready to be merged.

I am also working on a more comfortable object orientated cpp interface that internally calls the wrappers. But that is something for a later release.

MarkWieczorek commented on examples/cpp/makefile:
```diff
@@ -0,0 +1,22 @@
+OPTIM = -O0
+CXXFLAGS += -Wall -pedantic -std=c++17 $(OPTIM) 
+
+INCLUDE=-I ../../src
+
+LIB= -L ../../lib -lSHTOOLS -lgfortran \
+     -ldl -lfftw3 -lm -llapack -lblas \
```
I am ok with this for now, but we might want to consider getting rid of the relative locations of `../../src` and `../../lib` for the case where these have been moved elsewhere (this was an issue with macports/homebrew...). What I have done for the Fortran examples is the following:

In the main Makefile, I created targets like `fortran-tests`, and then called the Makefile in the examples folder using

```
@$(MAKE) -C $(FEXDIR) -f Makefile all F95="$(F95)" F95FLAGS="$(F95FLAGS)" LIBNAME="$(LIBNAME)" FFTW="$(FFTW)" LAPACK="$(LAPACK)" BLAS="$(BLAS)" LIBPATH="$(LIBPATH)" MODFLAG="$(MODFLAG)"
```

Its ugly, but it works. The good thing about this approach is that the main Makefile defines all of these variables once (some of which might be different for different compilers). The bad part is that the file `examples/cpp/Makefile` can't be used by itself (unless you set default values for all the variables, which are compiler and architecture dependent).

In any case, it would be useful to include a simple `cpp-tests` target in the main Makefile so that it could be easily run on Travis.

rainbowsend replied:
another way to deal with this could be creating 'machine files', for example, one called `linux-gfortran`:

```make
# Default gfortran flags
F95FLAGS ?= -m64 -fPIC -O3 -std=gnu -ffast-math
# -march=native
MODFLAG = -I$(MODPATH)
SYSMODFLAG = -I$(SYSMODPATH)
OPENMPFLAGS ?= -fopenmp
BLAS ?= -lblas
LAPACK ?= -llapack
```

include the machine file in the main makefile with

`include $(MAKE_MACHINE)`

and call make with

`MAKE_MACHINE=linux-gfortran make`

the sub makefiles can also include the corresponding file

MarkWieczorek commented:
I'm trying to compile this on macOS, but I am getting some weird errors. Do you have any idea as to what this means?

In any case, we should probably move the shtools.h file to the new `include` directory (renamed from `modules`).

```
make
g++-10  -Wall -pedantic -std=c++17 -O0  -I ../../src -c -o test.o test.cpp
g++-10  -o test test.o -L ../../lib -lSHTOOLS -lgfortran -ldl -L/usr/local/lib -lfftw3 -lm -framework Accelerate
Undefined symbols for architecture x86_64:
  "__ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE4sizeEv", referenced from:
      _main in test.o
  "__ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE5c_strEv", referenced from:
      _main in test.o
  "__ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7_M_dataEv", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_disposeEv", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE11_M_capacityEm", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_Alloc_hiderC1EPcRKS3_", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEC1IS3_EEPKcRKS3_ in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE13_M_local_dataEv", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEC1IS3_EEPKcRKS3_ in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE13_M_set_lengthEm", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE13_S_copy_charsEPcPKcS7_", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7_M_dataEPc", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_createERmm", referenced from:
      __ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag in test.o
  "__ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEED1Ev", referenced from:
      _main in test.o
  "__ZdlPvm", referenced from:
      __ZN9__gnu_cxx13new_allocatorIdE10deallocateEPdm in test.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make: *** [test] Error 1
```

rainbowsend commented:
That is a strange error. Did you try rebuilding libSHTOOLS.a

I have updated the cpp makefile  as suggested.

However, there is a problem with moving `shtools.h` to the include folder: The include folder is created by `make` and deleted by ` make clean`. But the header file is actual code and not some machine readable code produced by make.

MarkWieczorek commented:
Since shtools.h is a permanent file, I don't have a problem with removing the creation/deletion of the `include` folder.

MarkWieczorek commented:
I'm going to merge this now, but I'll put in a small PR making some minor changes to the makefile in a few minutes.

The problem with the above error is that my g++ was using the wrong libgfortran. I don't have time right now to figure out if this is my fault or not, but these sets of parameters worked using my homebrew installation of gcc-10:

```
g++-10 -o test test.o -L /Users/lunokhod/SphericalHarmonics/shtools-git/lib -lSHTOOLS -L /usr/local/lib/gcc/10/ -lgfortran -ldl -L/usr/local/lib -lfftw3 -lm -framework Accelerate
```
I didn't move the include file, but at some point we probably should.
